### PR TITLE
Replace the Go Batch implementation with one based on RocksDB WriteBatch.

### PR DESCRIPTION
**Pull request by petermattis**
Somewhat more involved than expected because the builtin RocksDB
WriteBatchWithIndex iterators do not handle merge records.

benchmark                         old ns/op     new ns/op     delta
BenchmarkMVCCBatch1Put10          6198          5223          -15.73%
BenchmarkMVCCBatch100Put10        6019          5586          -7.19%
BenchmarkMVCCBatch10000Put10      9324          5539          -40.59%
BenchmarkMVCCBatch100000Put10     13419         6242          -53.48%

benchmark                         old MB/s     new MB/s     speedup
BenchmarkMVCCBatch1Put10          1.61         1.91         1.19x
BenchmarkMVCCBatch100Put10        1.66         1.79         1.08x
BenchmarkMVCCBatch10000Put10      1.07         1.81         1.69x
BenchmarkMVCCBatch100000Put10     0.75         1.60         2.13x

### CODE REVIEW

File: storage/engine/db.cc [hunk 1]
```diff
@@ -739,6 +756,342 @@ class DBLogger : public rocksdb::Logger {
...
+struct Base {
+  virtual DBStatus Get(rocksdb::Slice key, DBString* value) = 0;
+};
+
+struct BaseIteratorImpl : public Base {
+  BaseIteratorImpl(rocksdb::Iterator* iter)
+      : base_(iter) {
+  }
+
+  virtual DBStatus Get(rocksdb::Slice key, DBString* value) {
```
Comment (spencerkimball) [inline storage/engine/db.cc]: It's a little strange that we ignore key here.

File: storage/engine/db.cc [hunk 2]
```diff
@@ -739,6 +756,342 @@ class DBLogger : public rocksdb::Logger {
   const bool enabled_;
 };
 
+struct Base {
```
Comment (spencerkimball) [inline storage/engine/db.cc]: Add a comment as to exactly what Base is. That it provides a common interface to accessing data from both an iterator and a batch.

File: storage/engine/db.cc (hunk 1 above)
Comment (petermattis) [inline storage/engine/db.cc]: Yes it is. I'll add a comment that we know `base_->key() == key`.

File: storage/engine/db.cc (hunk 2 above)
Comment (petermattis) [inline storage/engine/db.cc]: Ok, adding comment. And I'm going to rename this interfacae `Getter`.

**Comment by spencerkimball** [comment]
LGTM

**Comment by bdarnell** [comment]
LGTM

**Pull request merged**